### ISSUE TITLE
A toy viewer using viser

### ISSUE THREAD

**Issue by haoming-Yu** [JiahuiLei/MoSca#22]
Thanks for Lei's great work!
In order to enable free view observe, I use claude code to generate a viser scirpt that can run well on my pc.
You just need to first train the scene according to instructions, then `pip install viser==1.0.24`

`"""
Interactive free-viewpoint rendering of MoSca 4D scenes using Viser.

Usage:
    python free_view_viser.py \
        --cfg ./profile/demo/demo_fit.yaml \
        --logdir ./demo/duck/logs/demo_fit_native_add3_20260317_202146 \
        [--port 8080] [--host 0.0.0.0] [--render_scale 1.0]
"""

import os, sys, os.path as osp
import argparse
import time
import threading
import numpy as np
import torch
import viser
import viser.transforms as vtf

sys.path.append(osp.dirname(osp.abspath(__file__)))

from omegaconf import OmegaConf
from lib_moca.camera import MonocularCameras
from lib_mosca.dynamic_gs import DynSCFGaussian
from lib_mosca.static_gs import StaticGaussian
from lib_render.render_helper import render, GS_BACKEND

# ---------------------------------------------------------------------------
#  Model loading (reuses mosca_viz.py logic)
# ---------------------------------------------------------------------------

@torch.no_grad()
def load_scene(cfg_path: str, logdir: str, device: torch.device):
    cfg = OmegaConf.load(cfg_path)

    cams: MonocularCameras = MonocularCameras.load_from_ckpt(
        torch.load(osp.join(logdir, "photometric_cam.pth"), map_location=device)
    ).to(device)

    s_model: StaticGaussian = StaticGaussian.load_from_ckpt(
        torch.load(
            osp.join(logdir, f"photometric_s_model_{GS_BACKEND.lower()}.pth"),
            map_location=device,
        ),
        device=device,
    )

    d_model: DynSCFGaussian = DynSCFGaussian.load_from_ckpt(
        torch.load(
            osp.join(logdir, f"photometric_d_model_{GS_BACKEND.lower()}.pth"),
            map_location=device,
        ),
        device=device,
    )

    # Ensure all parameters are on the correct device
    # (load_from_ckpt may leave some params on CPU)
    cams.to(device).eval()
    s_model.to(device).eval()
    d_model.to(device).eval()

    # Bake skinning for faster forward pass
    d_model.set_inference_mode()

    return cfg, cams, s_model, d_model

# ---------------------------------------------------------------------------
#  Rendering helper
# ---------------------------------------------------------------------------

@torch.no_grad()
def render_frame(
    s_model,
    d_model,
    cams,
    T_cw: torch.Tensor,
    K: torch.Tensor,
    H: int,
    W: int,
    t: int,
    bg_color=(1.0, 1.0, 1.0),
    show_static: bool = True,
    show_dynamic: bool = True,
):
    """Render a single frame and return an (H, W, 3) uint8 numpy array."""
    gs5 = []
    if show_static:
        gs5.append(s_model())
    if show_dynamic:
        gs5.append(d_model(t))
    if len(gs5) == 0:
        return np.ones((H, W, 3), dtype=np.uint8) * 255

    render_dict = render(
        gs5,
        H,
        W,
        K=K,
        T_cw=T_cw,
        bg_color=list(bg_color),
    )
    rgb = torch.clamp(render_dict["rgb"].permute(1, 2, 0), 0.0, 1.0)
    return (rgb.detach().cpu().numpy() * 255).astype(np.uint8)

# ---------------------------------------------------------------------------
#  Viser ↔ MoSca coordinate helpers
# ---------------------------------------------------------------------------

def viser_cam_to_T_cw(camera: viser.CameraHandle) -> np.ndarray:
    """Convert a viser camera (world-from-camera SE3) to a 4×4 T_cw matrix
    in the MoSca convention (OpenCV: +X right, +Y down, +Z forward)."""
    # Viser uses OpenGL convention (+X right, +Y up, −Z forward)
    # We need to flip Y and Z axes to get OpenCV convention.
    R_wc = vtf.SO3(camera.wxyz).as_matrix()  # 3×3
    t_wc = np.array(camera.position)          # (3,)

    # OpenGL → OpenCV: flip Y and Z
    flip = np.diag([1.0, -1.0, -1.0])
    R_wc = R_wc @ flip
    T_wc = np.eye(4)
    T_wc[:3, :3] = R_wc
    T_wc[:3, 3] = t_wc
    T_cw = np.linalg.inv(T_wc)
    return T_cw

def T_cw_to_viser_pose(T_cw_np: np.ndarray):
    """Convert a MoSca T_cw (OpenCV) to viser camera (wxyz, position)."""
    T_wc = np.linalg.inv(T_cw_np)
    R_wc = T_wc[:3, :3]
    t_wc = T_wc[:3, 3]
    flip = np.diag([1.0, -1.0, -1.0])
    R_wc_gl = R_wc @ flip
    wxyz = vtf.SO3.from_matrix(R_wc_gl).wxyz
    return wxyz, t_wc

# ---------------------------------------------------------------------------
#  Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser("MoSca Viser Free-View Renderer")
    parser.add_argument("--cfg", type=str, required=True, help="Config yaml path")
    parser.add_argument("--logdir", type=str, required=True, help="Log directory")
    parser.add_argument("--port", type=int, default=8080)
    parser.add_argument("--host", type=str, default="0.0.0.0")
    parser.add_argument("--render_scale", type=float, default=1.0,
                        help="Down-scale factor for render resolution (0~1)")
    parser.add_argument("--fps", type=float, default=15.0,
                        help="Playback FPS for timeline")
    args = parser.parse_args()

    device = torch.device("cuda:0")
    print(f"Loading scene from {args.logdir} ...")
    cfg, cams, s_model, d_model = load_scene(args.cfg, args.logdir, device)
    T_total = cams.T
    H_orig = int(cams.default_H)
    W_orig = int(cams.default_W)
    print(f"Scene loaded: {T_total} frames, {H_orig}×{W_orig}, GS_BACKEND={GS_BACKEND}")

    # Render resolution
    scale = args.render_scale
    H_render = max(1, int(H_orig * scale))
    W_render = max(1, int(W_orig * scale))

    # ------------------------------------------------------------------
    #  Viser server
    # ------------------------------------------------------------------
    server = viser.ViserServer(host=args.host, port=args.port)
    print(f"Viser server started at http://{args.host}:{args.port}")

    # ---- GUI controls ----
    with server.gui.add_folder("Playback"):
        gui_playing = server.gui.add_checkbox("Play", initial_value=False)
        gui_timestep = server.gui.add_slider(
            "Timestep", min=0, max=T_total - 1, step=1, initial_value=0
        )
        gui_fps = server.gui.add_slider(
            "FPS", min=1.0, max=60.0, step=1.0, initial_value=args.fps
        )
        gui_loop = server.gui.add_checkbox("Loop", initial_value=True)

    with server.gui.add_folder("Rendering"):
        gui_show_static = server.gui.add_checkbox("Static GS", initial_value=True)
        gui_show_dynamic = server.gui.add_checkbox("Dynamic GS", initial_value=True)
        gui_bg_white = server.gui.add_checkbox("White Background", initial_value=True)
        gui_render_scale = server.gui.add_slider(
            "Render Scale", min=0.1, max=1.0, step=0.05, initial_value=scale
        )

    with server.gui.add_folder("Camera"):
        gui_snap_training = server.gui.add_button("Snap to Training View")
        gui_cam_info = server.gui.add_text("Camera Info", initial_value="", disabled=True)

    # ---- Add training camera frustums for reference ----
    with server.gui.add_folder("Visualisation"):
        gui_show_frustums = server.gui.add_checkbox("Show Training Cameras", initial_value=False)

    frustum_handles = []

    @gui_show_frustums.on_update
    def _toggle_frustums(event: viser.GuiEvent) -> None:
        nonlocal frustum_handles
        if gui_show_frustums.value:
            # Draw frustums for a subset of training cameras
            step = max(1, T_total // 20)
            for t_idx in range(0, T_total, step):
                T_cw_np = cams.T_cw(t_idx).detach().cpu().numpy()
                wxyz, position = T_cw_to_viser_pose(T_cw_np)
                fov_y = float(np.deg2rad(cams.fov[1]))
                h = server.scene.add_camera_frustum(
                    f"/frustums/cam_{t_idx:04d}",
                    fov=fov_y,
                    aspect=W_orig / H_orig,
                    scale=0.05,
                    color=(80, 180, 80),
                    wxyz=wxyz,
                    position=position,
                )
                frustum_handles.append(h)
        else:
            for h in frustum_handles:
                h.remove()
            frustum_handles.clear()

    # ---- Snap to training view ----
    @gui_snap_training.on_click
    def _snap_to_training(event: viser.GuiEvent) -> None:
        if event.client is None:
            return
        t = int(gui_timestep.value)
        T_cw_np = cams.T_cw(t).detach().cpu().numpy()
        wxyz, position = T_cw_to_viser_pose(T_cw_np)
        event.client.camera.wxyz = wxyz
        event.client.camera.position = position

    # ---- Set initial camera from frame 0 ----
    @server.on_client_connect
    def _on_connect(client: viser.ClientHandle) -> None:
        T_cw_np = cams.T_cw(0).detach().cpu().numpy()
        wxyz, position = T_cw_to_viser_pose(T_cw_np)
        client.camera.wxyz = wxyz
        client.camera.position = position
        # Set reasonable camera properties
        fov_y = float(np.deg2rad(cams.fov[1]))
        client.camera.fov = fov_y

    # ------------------------------------------------------------------
    #  Render loop (per client)
    # ------------------------------------------------------------------

    # Global playback state
    last_advance_time = time.time()

    def playback_tick():
        """Advance timestep if playing."""
        nonlocal last_advance_time
        if not gui_playing.value:
            last_advance_time = time.time()
            return
        now = time.time()
        dt = now - last_advance_time
        interval = 1.0 / gui_fps.value
        if dt >= interval:
            last_advance_time = now
            t = int(gui_timestep.value)
            t += 1
            if t >= T_total:
                if gui_loop.value:
                    t = 0
                else:
                    t = T_total - 1
                    gui_playing.value = False
            gui_timestep.value = t

    # Render & stream to each connected client
    while True:
        playback_tick()

        for client in server.get_clients().values():
            cam = client.camera

            # Compute camera matrices
            T_cw_np = viser_cam_to_T_cw(cam)
            T_cw_torch = torch.from_numpy(T_cw_np).float().to(device)

            # Use client viewport fov to build K
            fov_y = cam.fov  # radians
            cur_scale = gui_render_scale.value
            H_r = max(1, int(H_orig * cur_scale))
            W_r = max(1, int(W_orig * cur_scale))
            fy = H_r / (2.0 * np.tan(fov_y / 2.0))
            fx = fy  # square pixels
            cx = W_r / 2.0
            cy = H_r / 2.0
            K = torch.tensor([
                [fx, 0.0, cx],
                [0.0, fy, cy],
                [0.0, 0.0, 1.0],
            ], dtype=torch.float32, device=device)

            bg = (1.0, 1.0, 1.0) if gui_bg_white.value else (0.0, 0.0, 0.0)
            t = int(gui_timestep.value)

            img = render_frame(
                s_model, d_model, cams,
                T_cw=T_cw_torch,
                K=K,
                H=H_r,
                W=W_r,
                t=min(t, T_total - 1),
                bg_color=bg,
                show_static=gui_show_static.value,
                show_dynamic=gui_show_dynamic.value,
            )

            client.scene.set_background_image(img, format="jpeg")

            gui_cam_info.value = (
                f"t={t}/{T_total-1}  res={W_r}×{H_r}  "
                f"fov={np.rad2deg(fov_y):.1f}°"
            )

        time.sleep(0.01)

if __name__ == "__main__":
    main()
`
